src/union.ts

```diff
@@ -25,9 +25,85 @@ export class Union {
             return false;
         };
 
-        // Special cases for methods which need to join their results
-        this['readdir'] = (...args) => this.asyncJoinMethod("readdir", args);
-        this['readdirSync'] = (...args) => this.syncJoinMethod("readdirSync", args);
+        // special case for readdir which should union the results
+        this['readdir'] = (...args) => {
+            let lastarg = args.length - 1;
+            let cb = args[lastarg];
+            if(typeof cb !== 'function') {
+                cb = null;
+                lastarg++;
+            }
+
+            let lastError: IUnionFsError = null;
+            let result: Set<string> | null = null;
+            const iterate = (i = 0, error?: IUnionFsError) => {
+                if(error) {
+                    error.prev = lastError;
+                    lastError = error;
+                }
+
+                // Already tried all file systems, return the last error.
+                if(i >= this.fss.length) { // last one
+                    if(cb) {
+                        cb(error || Error('No file systems attached.'));
+                    };
+                    return;
+                }
+
+                // Replace `callback` with our intermediate function.
+                args[lastarg] = (err, resArg: string[] | Buffer[]) => {
+                    if(err) {
+                        return iterate(i + 1, err);
+                    }
+                    if(resArg) {
+                        result = result !== null ? result : new Set();
+                        
+                        // Convert all results to Strings to make sure that they're deduped
+                        for (const res of resArg) {
+                            result.add(String(res));
+                        }
+                    }
+
+                    if (i === this.fss.length - 1) {
+                        return cb(null, Array.from(result));
```

Comment:
I think on some systems `readdir` results are sorted alphabetically. Do you think we should do this here?